### PR TITLE
[IA-2829] n-1 approach for legacy images in dropdown menu 

### PR DESCRIPTION
**Pull request by Kfunes**
This code changes add logic to automate the computation of the legacy GATK and Bioconductor images. It defines the legacy image as the image before a major or minor version bump. 

Testing: 
The current versions in the dropdown menu are: 
GATK-2.0.0
Bioconductor-2.0.0
We tested a minor version bump for the GATK image. Here is `terra-docker-versions-new.json`: 
[
    {
        "updated": "2021-07-08", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-bioconductor:2.0.0", 
        "label": "R / Bioconductor: (Python 3.7.10, R 4.1.0, BiocVersion 3.13.1, tidyverse 1.3.1)", 
        "version": "2.0.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-bioconductor-2.0.0-versions.json", 
        "id": "terra-jupyter-bioconductor"
    }, 
    {
        "updated": "2021-06-29", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-hail:1.0.0", 
        "label": "Hail: (Python 3.7.10, Spark 2.4.5, hail 0.2.62)", 
        "version": "1.0.0", 
        "requiresSpark": true, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-hail-1.0.0-versions.json", 
        "id": "terra-jupyter-hail"
    }, 
    {
        "updated": "2021-06-29", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:1.0.0", 
        "label": "Python: (Python 3.7.10, pandas 1.2.4, scikit-learn 0.24.2)", 
        "version": "1.0.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-python-1.0.0-versions.json", 
        "id": "terra-jupyter-python"
    }, 
    {
        "updated": "", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.1.0", 
        "label": "Default: (GATK THIS IS A TEST, Python THIS IS A TEST, R THIS IS A TEST)", 
        "version": "2.1.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-gatk-2.1.0-versions.json", 
        "id": "terra-jupyter-gatk"
    }, 
    {
        "updated": "2021-06-25", 
        "image": "cumulusprod/pegasus-terra:1.4.0", 
        "label": "Pegasus (Pegasuspy 1.4.0, Python 3.7.10, harmony-pytorch 0.1.6, nmf-torch 0.1.1)", 
        "version": "1.4.0", 
        "isCommunity": true, 
        "requiresSpark": false, 
        "packages": "https://raw.githubusercontent.com/klarman-cell-observatory/cumulus/master/docker/pegasus-terra/1.4.0/pegasus-terra-1_4_0-versions.json", 
        "id": "Pegasus"
    }, 
    {
        "updated": "2021-06-24", 
        "image": "us.gcr.io/anvil-gcr-public/anvil-rstudio-bioconductor:3.13.2", 
        "label": "RStudio (R 4.1.0, Bioconductor 3.13.0, Python 3.8.5)", 
        "version": "3.13.2", 
        "requiresSpark": false, 
        "isRStudio": true, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/placeholder.json", 
        "id": "RStudio"
    }, 
    {
        "updated": "2021-07-20", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.1", 
        "label": "Default: (GATK 4.2.0.0, Python 3.7.10, R 4.1.0)", 
        "version": "2.0.1", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-gatk-2.0.1-versions.json", 
        "id": "terra-jupyter-gatk"
    }, 
    {
        "updated": "2021-05-05", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-bioconductor:1.0.14", 
        "label": "Legacy R / Bioconductor (R 4.0.5, Bioconductor 3.12, Python 3.8.5)", 
        "version": "1.0.14", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-bioconductor-1.0.14-versions.json", 
        "id": "terra-jupyter-bioconductor_legacy"
    }
]

We tested a minor version bump for the Bioconductor image. Here is `terra-docker-versions-new.json`: 
[
    {
        "updated": "2021-01-26", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-bioconductor:2.1.0", 
        "label": "R / Bioconductor: (Python THIS IS A TEST, R THIS IS A TEST, BiocVersion THIS IS A TEST, tidyverse THIS IS A TEST)", 
        "version": "2.1.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-bioconductor-2.1.0-versions.json", 
        "id": "terra-jupyter-bioconductor"
    }, 
    {
        "updated": "2021-06-29", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-hail:1.0.0", 
        "label": "Hail: (Python 3.7.10, Spark 2.4.5, hail 0.2.62)", 
        "version": "1.0.0", 
        "requiresSpark": true, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-hail-1.0.0-versions.json", 
        "id": "terra-jupyter-hail"
    }, 
    {
        "updated": "2021-06-29", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:1.0.0", 
        "label": "Python: (Python 3.7.10, pandas 1.2.4, scikit-learn 0.24.2)", 
        "version": "1.0.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-python-1.0.0-versions.json", 
        "id": "terra-jupyter-python"
    }, 
    {
        "updated": "2021-07-20", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:2.0.1", 
        "label": "Default: (GATK 4.2.0.0, Python 3.7.10, R 4.1.0)", 
        "version": "2.0.1", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-gatk-2.0.1-versions.json", 
        "id": "terra-jupyter-gatk"
    }, 
    {
        "updated": "2021-06-25", 
        "image": "cumulusprod/pegasus-terra:1.4.0", 
        "label": "Pegasus (Pegasuspy 1.4.0, Python 3.7.10, harmony-pytorch 0.1.6, nmf-torch 0.1.1)", 
        "version": "1.4.0", 
        "isCommunity": true, 
        "requiresSpark": false, 
        "packages": "https://raw.githubusercontent.com/klarman-cell-observatory/cumulus/master/docker/pegasus-terra/1.4.0/pegasus-terra-1_4_0-versions.json", 
        "id": "Pegasus"
    }, 
    {
        "updated": "2021-06-24", 
        "image": "us.gcr.io/anvil-gcr-public/anvil-rstudio-bioconductor:3.13.2", 
        "label": "RStudio (R 4.1.0, Bioconductor 3.13.0, Python 3.8.5)", 
        "version": "3.13.2", 
        "requiresSpark": false, 
        "isRStudio": true, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/placeholder.json", 
        "id": "RStudio"
    }, 
    {
        "updated": "2020-06-29", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:1.1.3", 
        "label": "Default: (GATK 4.1.4.1, Python 3.7.8, R 4.0.2)", 
        "version": "1.1.3", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-gatk-1.1.3-versions.json", 
        "id": "terra-jupyter-gatk_legacy"
    }, 
    {
        "updated": "2021-07-08", 
        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-bioconductor:2.0.0", 
        "label": "R / Bioconductor: (Python 3.7.10, R 4.1.0, BiocVersion 3.13.1, tidyverse 1.3.1)", 
        "version": "2.0.0", 
        "requiresSpark": false, 
        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-bioconductor-2.0.0-versions.json", 
        "id": "terra-jupyter-bioconductor"
    }
]

We also tested a major version bump for GATK and then Bioconductor. We also tested a minor version bump for Bioconductor and a major version bump for GATK at the same time. We also tested _no_ version bump for either image and our code worked as we expected it to.

### CODE REVIEW

**Review by xuscrbw — APPROVED** [review]
LGTM. Comment on the new line. Also there's a comment on ln 61 "this is a test" I think might need to be removed, but not seeing anything.

File: config/legacy_static_images.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (Qi77Qi) [inline config/legacy_static_images.json:1]: looks like the legacy image json will be generated in the script...do we still need this file?

**Review by jdcanas — APPROVED** [review]
:shipit:

File: config/legacy_static_images.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    },
+  {
+        "updated": "2020-06-29", 
+        "image": "us.gcr.io/broad-dsp-gcr-public/terra-jupyter-gatk:1.1.3", 
+        "label": "Default: (GATK 4.1.4.1, Python 3.7.8, R 4.0.2)", 
+        "version": "1.1.3", 
+        "requiresSpark": false, 
+        "packages": "https://storage.googleapis.com/terra-docker-image-documentation/terra-jupyter-gatk-1.1.3-versions.json", 
+        "id": "terra-jupyter-gatk_legacy"
+    }]
```
[the remark this answers is not in the document]
Comment (Kfunes) [reply config/legacy_static_images.json]: Just pushed these changes

File: config/legacy_static_images.json (hunk 1 above)
Comment (AdryanneG) [reply config/legacy_static_images.json]: We are keeping this for now because the way the code works is that the image before a major or minor version bump will become the new legacy. The first time this script runs there will be no major or minor version bump so the static images are the legacy. This can be removed once there is a major or minor version bump for both the Bioconductor and GATK images.

**Review by rtitle — APPROVED** [review]
Looks good to me! :shipit:
